infer/tests/provider_test.go

```diff
@@ -581,23 +581,23 @@ func providerOpts(config infer.InferredConfig) infer.Options {
 	return infer.Options{
 		Config: config,
 		Resources: []infer.InferredResource{
-			infer.Resource[*Echo](),
-			infer.Resource[*Wired](),
-			infer.Resource[*WiredPlus](),
-			infer.Resource[*Increment](),
-			infer.Resource[*WithDefaults](),
-			infer.Resource[*ReadEnv](),
-			infer.Resource[*Recursive](),
-			infer.Resource[*ReadConfig](),
-			infer.Resource[*ReadConfigCustom](),
-			infer.Resource[*CustomCheckNoDefaults](),
+			infer.Resource(&Echo{}),
+			infer.Resource(&Wired{}),
+			infer.Resource(&WiredPlus{}),
+			infer.Resource(&Increment{}),
+			infer.Resource(&WithDefaults{}),
+			infer.Resource(&ReadEnv{}),
+			infer.Resource(&Recursive{}),
+			infer.Resource(&ReadConfig{}),
+			infer.Resource(&ReadConfigCustom{}),
+			infer.Resource(&CustomCheckNoDefaults{}),
 		},
 		Components: []infer.InferredComponent{
 			infer.Component(NewRandomComponent),
 			infer.Component(NewReadConfigComponent),
```

Comment:
I would consider it a blocker if we don't solve for component creation.  Hopefully the functional style can be a convenience that is also supported.

Some suggestions:
```go
Components: []infer.InferredComponent{
  infer.ComponentF(NewRandomComponent),
  infer.Component(&RandomComponent{}),
  infer.ComponentResource(&RandomComponent{}),
}
```

Reply:
I was thinking the same, I like `ComponentF`.

Reply:
As discussed this morning, the goal of allowing users to re-use existing components is still possible with the old API:

```go
func (r *RandomLogin) Construct(ctx *pulumi.Context, name string, typ string, args RandomLoginArgs, opts pulumi.ResourceOption) (*RandomLogin, error) {
    comp, err := NewRandomLogin(ctx, name, args, opts) /* existing code */
    if err != nil {
        return nil, err
    }
    return ctx.RegisterComponentResource(typ, name, comp, opts)
}
```

The PR includes `ComponentF` so users can still pass `NewRandomLogin` directly, but this prevents the user from benefiting from any of the framework's control over the schema (e.g., aliasing the component, deprecating it, etc.).

Reply:
I believe that you could still do all of that already by hanging the methods on the provider struct produced by the function.

Reply:
On @mikhailshilkov's point:
> The one explicit design goal of the Pulumi Components project was to keep the API of component definition of a "local" component and a "remote" component as close as possible, ideally exactly the same.

I have a concern that this goal was not achieved, if it is about the compatibility of a given component function as a local and remote component.  I say this because the user is now expected to call `p.GetTypeToken` within the function body, and this means that the function is effectively not a local component. Meanwhile, posing as a local component leaves you thread-bare when it comes to implementing advanced functionality, like a Call into the component.

Reply:
> the user is now expected to call p.GetTypeToken within the function body

This is not true. Our docs show it hard-coded, same as in other languages and local components.

Reply:
@mikhailshilkov thanks, I've been curious about this point, and will take steps to deprecate GetTypeToken.